File: app-service-comments/src/test/java/com/github/arhor/aws/graphql/federation/comments/service/impl/UserRepresentationServiceImplTest.java
package com.github.arhor.aws.graphql.federation.comments.service.impl;

import com.github.arhor.aws.graphql.federation.comments.data.model.UserRepresentation;
import com.github.arhor.aws.graphql.federation.comments.data.model.UserRepresentation.UserFeature;
import com.github.arhor.aws.graphql.federation.comments.data.model.UserRepresentation.UserFeatures;
import com.github.arhor.aws.graphql.federation.comments.data.repository.UserRepresentationRepository;
import com.github.arhor.aws.graphql.federation.comments.generated.graphql.DgsConstants.USER;
import com.github.arhor.aws.graphql.federation.comments.generated.graphql.types.User;
import com.github.arhor.aws.graphql.federation.common.exception.EntityConditionException;
import com.github.arhor.aws.graphql.federation.common.exception.EntityNotFoundException;
import com.github.arhor.aws.graphql.federation.common.exception.Operation;
import com.github.arhor.aws.graphql.federation.starter.testing.ConstantsKt;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.EnumSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;

import static com.github.arhor.aws.graphql.federation.starter.testing.MockitoExtKt.withFirstArg;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.catchException;
import static org.assertj.core.api.Assertions.from;
import static org.assertj.core.api.InstanceOfAssertFactories.type;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;
import static org.mockito.Mockito.mock;

class UserRepresentationServiceImplTest {

    private static final UUID USER_ID = ConstantsKt.getTEST_1_UUID_VAL();

    private final UserRepresentationRepository userRepository = mock();

    private final UserRepresentationServiceImpl userService = new UserRepresentationServiceImpl(
        userRepository
    );

    @Nested
    @DisplayName("Method findUsersRepresentationsInBatch")
    class FindUserRepresentationTest {
        @Test
        void should_return_expected_user_when_it_exists_by_id() {
            // Given
            final var userRepresentation =
                UserRepresentation.builder()
                    .id(USER_ID)
                    .build();

            final var expectedResult = Map.of(
                USER_ID,
                User.newBuilder()
                    .id(USER_ID)
                    .commentsDisabled(false)
                    .build()
            );
            final var expectedUserIds = Set.of(USER_ID);

            given(userRepository.findAllById(any()))
                .willReturn(List.of(userRepresentation));

            // When
            final var result = userService.findUsersRepresentationsInBatch(expectedUserIds);

            // Then
            then(userRepository)
                .should()
                .findAllById(expectedUserIds);

            then(userRepository)
                .shouldHaveNoMoreInteractions();

            assertThat(result)
                .isNotNull()
                .isEqualTo(expectedResult);
        }

        @Test
        void should_return_user_with_commentsOperable_false_when_user_does_not_exist_by_id() {
            // Given
            final var expectedResult = Map.of(
                USER_ID,
                User.newBuilder()
                    .id(USER_ID)
                    .build()
            );
            final var expectedUserIds = Set.of(USER_ID);

            given(userRepository.findAllById(any()))
                .willReturn(Collections.emptyList());

            // When
            final var result = userService.findUsersRepresentationsInBatch(expectedUserIds);

            // Then
            then(userRepository)
                .should()
                .findAllById(expectedUserIds);

            then(userRepository)
                .shouldHaveNoMoreInteractions();

            assertThat(result)
                .isNotNull()
                .isEqualTo(expectedResult);
        }
    }

    @Nested
    @DisplayName("Method toggleUserComments")
    class SwitchUserCommentsTest {
        @Test
        void should_disable_comments_for_a_given_post_when_they_were_enabled() {
            // Given
            final var user =
                UserRepresentation.builder()
                    .id(USER_ID)
                    .build();

            given(userRepository.findById(any()))
                .willReturn(Optional.of(user));

            given(userRepository.save(any()))
                .willAnswer(withFirstArg());

            // When
            final var result = userService.toggleUserComments(USER_ID);

            // Then
            then(userRepository)
                .should()
                .findById(USER_ID);

            then(userRepository)
                .should()
                .save(user.toBuilder().features(user.features().plus(UserFeature.COMMENTS_DISABLED)).build());

            assertThat(result)
                .isFalse();
        }

        @Test
        void should_enable_comments_for_a_given_post_when_they_were_disabled() {
            // Given
            final var user =
                UserRepresentation.builder()
                    .id(USER_ID)
                    .features(new UserFeatures(EnumSet.of(UserFeature.COMMENTS_DISABLED)))
                    .build();

            given(userRepository.findById(any()))
                .willReturn(Optional.of(user));

            given(userRepository.save(any()))
                .willAnswer(withFirstArg());

            // When
            final var result = userService.toggleUserComments(USER_ID);

            // Then
            then(userRepository)
                .should()
                .findById(USER_ID);

            then(userRepository)
                .should()
                .save(user.toBuilder().features(user.features().minus(UserFeature.COMMENTS_DISABLED)).build());

            assertThat(result)
                .isTrue();
        }

        @Test
        void should_throw_EntityNotFoundException_when_there_is_no_user_found_by_the_input_id() {
            // Given
            given(userRepository.findById(any()))
                .willReturn(Optional.empty());

            // When
            final var result = catchException(() -> userService.toggleUserComments(USER_ID));

            // Then
            then(userRepository)
                .should()
                .findById(USER_ID);

            assertThat(result)
                .asInstanceOf(type(EntityNotFoundException.class))
                .returns(USER.TYPE_NAME, from(EntityConditionException::getEntity))
                .returns(USER.Id + " = " + USER_ID, from(EntityConditionException::getCondition))
                .returns(Operation.UPDATE, from(EntityConditionException::getOperation));
        }
    }
}

File: app-service-posts/src/test/kotlin/com/github/arhor/aws/graphql/federation/posts/service/impl/UserRepresentationServiceImplTest.kt
package com.github.arhor.aws.graphql.federation.posts.service.impl

import com.github.arhor.aws.graphql.federation.common.exception.EntityNotFoundException
import com.github.arhor.aws.graphql.federation.common.exception.Operation
import com.github.arhor.aws.graphql.federation.posts.data.model.UserRepresentation
import com.github.arhor.aws.graphql.federation.posts.data.model.UserRepresentation.UserFeature
import com.github.arhor.aws.graphql.federation.posts.data.model.UserRepresentation.UserFeatures
import com.github.arhor.aws.graphql.federation.posts.data.repository.UserRepresentationRepository
import com.github.arhor.aws.graphql.federation.posts.generated.graphql.DgsConstants.USER
import com.github.arhor.aws.graphql.federation.posts.generated.graphql.types.User
import com.github.arhor.aws.graphql.federation.posts.service.UserRepresentationService
import com.github.arhor.aws.graphql.federation.starter.testing.OMNI_UUID_VAL
import com.github.arhor.aws.graphql.federation.starter.testing.ZERO_UUID_VAL
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import io.mockk.junit5.MockKExtension
import io.mockk.just
import io.mockk.runs
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchException
import org.assertj.core.api.Assertions.from
import org.assertj.core.api.InstanceOfAssertFactories.type
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.cache.CacheType
import org.springframework.boot.test.autoconfigure.core.AutoConfigureCache
import org.springframework.cache.annotation.EnableCaching
import org.springframework.test.context.junit.jupiter.SpringJUnitConfig
import java.util.Optional

@ExtendWith(MockKExtension::class)
@EnableCaching
@SpringJUnitConfig(UserRepresentationServiceImpl::class)
@AutoConfigureCache(cacheProvider = CacheType.CAFFEINE)
class UserRepresentationServiceImplTest {

    @MockkBean
    private lateinit var userRepository: UserRepresentationRepository

    @Autowired
    private lateinit var userService: UserRepresentationService

    @Test
    fun `findUsersRepresentationsInBatch should return expected user when it exists by id`() {
        // Given
        val userRepresentation = UserRepresentation(USER_ID)
        val expectedResult = mapOf(USER_ID to User(id = userRepresentation.id, postsDisabled = false))
        val expectedUserIds = setOf(USER_ID)

        every { userRepository.findAllById(any()) } returns listOf(userRepresentation)

        // When
        val result = userService.findUsersRepresentationsInBatch(expectedUserIds)

        // Then
        verify(exactly = 1) { userRepository.findAllById(expectedUserIds) }

        assertThat(result)
            .isNotNull()
            .isEqualTo(expectedResult)
    }

    @Test
    fun `findUsersRepresentationsInBatch should return user with postsOperable false when user does not exist by id`() {
        // Given
        val user = User(id = USER_ID)
        val expectedResult = mapOf(user.id to user)
        val expectedUserIds = setOf(user.id)

        every { userRepository.findAllById(any()) } returns emptyList()

        // When
        val result = userService.findUsersRepresentationsInBatch(expectedUserIds)

        // Then
        verify(exactly = 1) { userRepository.findAllById(expectedUserIds) }

        assertThat(result)
            .isNotNull()
            .isEqualTo(expectedResult)
    }

    @Test
    fun `createUserRepresentation should call userRepository save only once with the same user id`() {
        // Given
        val expectedUser = UserRepresentation(
            id = USER_ID,
            shouldBePersisted = true,
        )

        every { userRepository.save(any()) } answers { firstArg() }

        // When
        for (i in 0..2) {
            userService.createUserRepresentation(USER_ID, IDEMPOTENCY_KEY)
        }

        // Then
        verify(exactly = 1) { userRepository.save(expectedUser) }
    }

    @Test
    fun `deleteUserRepresentation should call userRepository deleteById only once with the same user id`() {
        // Given
        every { userRepository.deleteById(any()) } just runs

        // When
        for (i in 0..2) {
            userService.deleteUserRepresentation(USER_ID, IDEMPOTENCY_KEY)
        }

        // Then
        verify(exactly = 1) { userRepository.deleteById(USER_ID) }
    }

    @Test
    fun `toggleUserPosts should call userRepository#save when there is update applied to the user`() {
        // Given
        val user = UserRepresentation(id = USER_ID)

        every { userRepository.findById(any()) } returns Optional.of(user)
        every { userRepository.save(any()) } answers { firstArg() }

        // When
        val result = userService.toggleUserPosts(USER_ID)

        // Then
        verify(exactly = 1) { userRepository.findById(USER_ID) }
        verify(exactly = 1) { userRepository.save(user.copy(features = user.features + UserFeature.POSTS_DISABLED)) }

        assertThat(result)
            .isFalse()
    }

    @Test
    fun `toggleUserPosts should not call userRepository#save when there is no update applied to the user`() {
        // Given
        val user = UserRepresentation(id = USER_ID, features = UserFeatures(UserFeature.POSTS_DISABLED))

        every { userRepository.findById(any()) } returns Optional.of(user)
        every { userRepository.save(any()) } answers { firstArg() }

        // When
        val result = userService.toggleUserPosts(USER_ID)

        // Then
        verify(exactly = 1) { userRepository.findById(USER_ID) }
        verify(exactly = 1) { userRepository.save(user.copy(features = user.features - UserFeature.POSTS_DISABLED)) }

        assertThat(result)
            .isTrue()
    }

    @Test
    fun `toggleUserPosts should throw EntityNotFoundException when there is no user found by the input id`() {
        // Given
        every { userRepository.findById(any()) } returns Optional.empty()

        // When
        val result = catchException { userService.toggleUserPosts(USER_ID) }

        // Then
        verify(exactly = 1) { userRepository.findById(USER_ID) }

        assertThat(result)
            .asInstanceOf(type(EntityNotFoundException::class.java))
            .returns(USER.TYPE_NAME, from { it.entity })
            .returns("${USER.Id} = $USER_ID", from { it.condition })
            .returns(Operation.UPDATE, from { it.operation })
    }

    companion object {
        private val USER_ID = ZERO_UUID_VAL
        private val IDEMPOTENCY_KEY = OMNI_UUID_VAL
    }
}
